tests/unit/parsers/test.manifestjson.js

```diff
@@ -1377,6 +1377,56 @@ describe('ManifestJSONParser', () => {
     }
   });
 
+  describe('optional only permissions', () => {
+    const OPTIONAL_ONLY_PERMISSIONS = ['userScripts', 'trialML'];
+
+    it.each(OPTIONAL_ONLY_PERMISSIONS)(
+      'should warn on optional-only permission "%s" requested as non-optional',
+      (permName) => {
+        const linter = new Linter({ _: ['bar'] });
+        const json = validManifestJSON({
+          permissions: ['tabs', permName],
+          browser_specific_settings: { gecko: { strict_max_version: '135.0' } },
+        });
+        const manifestJSONParser = new ManifestJSONParser(
+          json,
+          linter.collector
+        );
+        expect(manifestJSONParser.collector.warnings).toEqual(
+          expect.arrayContaining([
+            expect.objectContaining({
+              code: 'MANIFEST_PERMISSIONS',
+              instancePath: '/permissions/1',
+              message: expect.stringMatching(
+                new RegExp(`Invalid permissions "${permName}"`)
+              ),
+            }),
+          ])
+        );
```

Comment:
note: @willdurand this is currently covering the behavior expected at the moment without a more specific error code, I'd be leaning towards filing a separate followup to consider introducing a separate error code and a better error message linking a doc page with more precise guidance about how optional only permissions should be used by extensions developers.